lib/charms/vault_k8s/v0/vault_autounseal.py

```diff
@@ -0,0 +1,371 @@
+        relation = self.framework.model.get_relation(
+            snapshot["relation_name"], snapshot["relation_id"]
+        )
+        if relation is None:
+            raise ValueError(
+                f"Unable to restore {self}: relation {snapshot['relation_name']} (id={snapshot['relation_id']}) not found."
+            )
+        self.relation = relation
+
+
+class VaultAutounsealDestroy(ops.EventBase):
+    """Event emitted when Vault autounseal configuration for a relation should be destroyed."""
+
+    def __init__(self, handle: ops.Handle, relation: model.Relation):
+        super().__init__(handle)
+        self.relation = relation
+
+    def snapshot(self) -> Dict[str, Any]:
+        """Return snapshot data that should be persisted."""
+        return dict(
+            super().snapshot(),
+            relation_id=self.relation.id,
+            relation_name=self.relation.name,
+        )
+
+    def restore(self, snapshot: Dict[str, Any]) -> None:
+        """Restore the event from a snapshot."""
+        super().restore(snapshot)
+        relation = self.framework.model.get_relation(
+            snapshot["relation_name"], snapshot["relation_id"]
+        )
+        if relation is None:
+            raise ValueError(
+                f"Unable to restore {self}: relation {snapshot['relation_name']} (id={snapshot['relation_id']}) not found."
+            )
+        self.relation = relation
+
+
+class VaultAutounsealProvidesEvents(ops.ObjectEvents):
+    """Events raised by the vault-autounseal relation on the provider side."""
+
+    vault_autounseal_initialize = ops.EventSource(VaultAutounsealInitialize)
+    vault_autounseal_destroy = ops.EventSource(VaultAutounsealDestroy)
+
+
+class VaultAutounsealRequireEvents(ops.ObjectEvents):
+    """Events raised by the vault-autounseal relation on the requirer side."""
+
+    vault_autounseal_details_ready = ops.EventSource(VaultAutounsealDetailsReadyEvent)
+
+
+class VaultAutounsealProvides(ops.Object):
+    """Manages the vault-autounseal relation from the provider side."""
+
+    on: VaultAutounsealProvidesEvents = VaultAutounsealProvidesEvents()  # type: ignore
+
+    def __init__(self, charm: ops.CharmBase, relation_name: str):
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+
+        self.framework.observe(
+            self.charm.on[relation_name].relation_created, self._on_relation_created
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_broken, self._on_relation_broken
+        )
+
+    def _on_relation_created(self, event: ops.RelationCreatedEvent) -> None:
+        self.on.vault_autounseal_initialize.emit(relation=event.relation)
+
+    def _on_relation_broken(self, event: ops.RelationBrokenEvent) -> None:
+        self.on.vault_autounseal_destroy.emit(relation=event.relation)
+
+    def set_vault_url(self, relation: ops.Relation, address: str) -> None:
```

Comment:
Is there value in having a bunch of independent sets instead of consolidating them together?

Reply:
I don't have a strong opinion here, I just followed the same pattern as vault_kv.

Reply:
I don't know that `vault_kv` is the best of our Python library, I wouldn't take too much from it :laughing:

Reply:
fixed in https://github.com/canonical/vault-k8s-operator/pull/346/commits/0651467cd3a75065ba0f04391194ceef96844526